### PR TITLE
update XMLwrapper for mxml 3.0

### PR DESCRIPTION
**Pull request by michiboo**

### CODE REVIEW

File: src/Misc/XMLwrapper.cpp [hunk 1]
```diff
@@ -528,39 +528,39 @@ void XMLwrapper::getparstr(const string &name, char *par, int maxstrlen) const
...
         return;
     }
 }
 
 string XMLwrapper::getparstr(const string &name,
                              const std::string &defaultpar) const
 {
-    const mxml_node_t *tmp = mxmlFindElement(node,
+    mxml_node_t *tmp = mxmlFindElement(node,
                                              node,
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: Changing this line results in the old alignment of other args (node, "string", "name", name.c_str()) being incorrect as they are now indented too far for the new code.

File: src/Misc/XMLwrapper.cpp [hunk 2]
```diff
@@ -670,15 +670,26 @@ void XMLwrapper::add(const XmlNode &node_)
...
+            char *name;
+            for (int i = 0; i < count; ++i)
+            {
+                n[name] = mxmlElementGetAttrByIndex(current, i, &name);
+            }
+        #else
             auto elm = current->value.element;
             XmlNode n(elm.name);
-            for(int i=0; i<elm.num_attrs; ++i) {
+            for (int i = 0; i < elm.num_attrs; ++i)
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: Unnecessary whitespace change

File: src/Misc/XMLwrapper.cpp [hunk 3]
```diff
@@ -670,15 +670,26 @@ void XMLwrapper::add(const XmlNode &node_)
 std::vector<XmlNode> XMLwrapper::getBranch(void) const
 {
     std::vector<XmlNode> res;
-    mxml_node_t *current = node->child;
+    mxml_node_t *current = mxmlGetFirstChild(node);
     while(current) {
-        if(current->type == MXML_ELEMENT) {
+        if (mxmlGetType(current) == MXML_ELEMENT) {
+        #if MXML_MAJOR_VERSION == 3
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: After double checking with my system version of mxml (a rather old one). It appears that not all versions have MXML_MAJOR_VERSION defined. For particularly old versions of mxml, can you use the new code if MXML_MAJOR_VERSION is defined and is equal to 3?

Also, by convention \#if \#else and \#endif are placed with no indention.

File: src/Misc/XMLwrapper.cpp [hunk 4]
```diff
@@ -519,7 +519,7 @@ int XMLwrapper::getparbool(const string &name, int defaultpar) const
 void XMLwrapper::getparstr(const string &name, char *par, int maxstrlen) const
 {
     ZERO(par, maxstrlen);
-    const mxml_node_t *tmp = mxmlFindElement(node,
+    mxml_node_t *tmp = mxmlFindElement(node,
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: Removal of 'const' breaks the alignment of remaining arguments

File: src/Misc/XMLwrapper.cpp [hunk 5]
```diff
@@ -670,15 +670,26 @@ void XMLwrapper::add(const XmlNode &node_)
...
-    mxml_node_t *current = node->child;
+    mxml_node_t *current = mxmlGetFirstChild(node);
     while(current) {
-        if(current->type == MXML_ELEMENT) {
+        if (mxmlGetType(current) == MXML_ELEMENT) {
+        #if MXML_MAJOR_VERSION == 3
+            XmlNode n(mxmlGetElement(current));
+            int count = mxmlElementGetAttrCount(current);
+            char *name;
+            for (int i = 0; i < count; ++i)
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: The common style is 'for() {' not 'for ()\n{'

**Comment by fundamental** [comment]
Looks good besides a few minor style changes. I'd also recommend https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html for general commit message styling, though that's not a strict rule.

Ping me once the changes are done and I'll merge the change in. I was seeing some discussion elsewhere about the security implications for the MXML 3.0 release, so once this is merged in then that will likely trigger a release.

**Comment by michiboo** [comment]
Hi I have updated the changes as you requested, but the build failed on Travis CI
due to The command "wget http://downloads.sourceforge.net/liblo/liblo-0.28.tar.gz" failed and exited with 4 during .

Can you please rerun it?

Thanks!

**Comment by fundamental** [comment]
Per the most recent commit message, I would change:

```
 ADDED MXML 3.0 COMPATIBILITY, added check for MXML version and compatibility for MXML3.0

- added #ifdef for MXML_MAJOR_VERSION in XMLwrapper.h for version that don't have it defined
- replace internal structure with MXML 3.0  API
```

To

```
Add MXML 3.0 Compatability

Add check for MXML version and compatibility for MXML 3.0:
- Add #ifdef for MXML_MAJOR_VERSION in XMLwrapper.h for version that 
  don't have it defined
- Replace internal structure with MXML 3.0  API
```

That should fix capitalization issues, present vs past tense issues, and line length issues.

File: src/Misc/XMLwrapper.cpp [hunk 6]
```diff
@@ -672,24 +672,22 @@ std::vector<XmlNode> XMLwrapper::getBranch(void) const
     std::vector<XmlNode> res;
     mxml_node_t *current = mxmlGetFirstChild(node);
     while(current) {
-        if (mxmlGetType(current) == MXML_ELEMENT) {
-        #if MXML_MAJOR_VERSION == 3
+        if(mxmlGetType(current) == MXML_ELEMENT) {
+#if MXML_MAJOR_VERSION == 3
```
Comment (fundamental) [inline src/Misc/XMLwrapper.cpp]: Trailing whitespace

**Comment by fundamental** [comment]
I've fixed the travis build on the master branch. Since this was passing and the changes since have been whitespace only it should be fine if the last build doesn't register correctly on travis.
I'd recommend a "git commit --amend" to correct the commit message, but otherwise everything is looking good and I should merge this afternoon.

**Comment by michiboo** [comment]
Hi I have updated the commit message and required change!

**Comment by fundamental** [comment]
Thanks for the fix :+1:
